### PR TITLE
fix crafting cycles

### LINKED ISSUE #4014: StringBuilder StackOverflowError Crashing Game
I left my colony to do some exploring while my builders are working and when I tried coming back to base my server crashed. Upon restarting and joining again, it still crashes the server when I get near my colony after the chat shows the builders resume their work.

https://pastebin.com/ZHpG7Lm3

### Minecolonies Versions

Tried both of these:

- minecolonies-1.12.2-0.10.488-RELEASE-universal.jar
- minecolonies-1.12.2-0.10.499-ALPHA-universal.jar

### Expected behavior

No crash to occur.

### Actual behaviour

Crashes game.

### Steps to reproduce the problem

No idea tbh. It looks like some issue with the warehouse resolving, but I don't know what specifically is being requested to cause the issue since it crashes before I can get to the builders and check. It's a lvl3 nordic warehouse. I've got one lvl3 nordic builder working on upgrading my nordic town hall to lvl3. The other builder is a lvl2 dwarven builder who is currently upgrading himself to lvl3.

The last time I was able to check on them, the lvl2 builder was trying to get 88 birch planks from the warehouse and some birch stairs from my sawmill crafter. No idea what the other might have been working on.

Hopefully the crash log illuminates what could be the cause.

### PR DESCRIPTION
**Pull request by Raycoms**
Closes #4014

# Changes proposed in this pull request:
- Fix crashes and crafting cycle detection.

This fixes the crafting cycle detection. It does this by tracking the different steps and checking if there is repetition inbetween.

Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/core/AbstractCraftingRequestResolver.java [hunk 1]
```diff
@@ -107,33 +109,46 @@ public boolean canResolveForBuilding(@NotNull final IRequestManager manager, @No
...
     protected boolean createsCraftingCycle(
-            @NotNull final IRequestManager manager,
-            @NotNull final IRequest<?> request,
-            @NotNull final IRequest<? extends IDeliverable> target,
-            final int count)
+      @NotNull final IRequestManager manager,
+      @NotNull final IRequest<?> request,
+      @NotNull final IRequest<? extends IDeliverable> target,
+      final int count,
+      final List reqs)
```
Comment (marchermans) [inline src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/core/AbstractCraftingRequestResolver.java]: None generic list!

**Review by marchermans — CHANGES_REQUESTED** [review]

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/core/AbstractCraftingRequestResolver.java (hunk 1 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/core/AbstractCraftingRequestResolver.java]: We got a generic request, I can only put them in generic lists =D

Comment (marchermans) [reply src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/core/AbstractCraftingRequestResolver.java]: What are you talking about?
You added the parameter, you know the types you put in it, specify the damn types

Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/core/AbstractCraftingRequestResolver.java]: Weird, Intellij wasn't giving me the type of this, I had to manually search it out and put it in...

Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/core/AbstractCraftingRequestResolver.java]: Intellij told me it was generic "?"

Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/core/AbstractCraftingRequestResolver.java]: I pushed the fix anyway

**Review by marchermans — APPROVED** [review]

**Pull request merged**